File: letter_combinations/jiadai.py
#jiadai c++ Your submission beats 99.40% Submissions!
class Solution:
    '''
     * @param digits: A digital string
     * @return: all posible letter combinations
    '''
    def letterCombinations(self, digits):
        #write your code here
        result = []
        if  len(digits) == 0:
            return result

        dict1 = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        self.dfs(digits, 0, "", dict1, result)
        return result

    def dfs(self, digits, start, path, dict1, result):
        if start == len(digits):
            result.append(path)
            return

        for c in dict1[ord(digits[start]) - ord('0')]:
            self.dfs(digits, start + 1, path + c, dict1, result)

File: string-permutation/jiadai.java
//jiadai, Your submission beats 91.80% Submissions!
public class Solution {
    /**
     * @param A: a string
     * @param B: a string
     * @return: a boolean
     */
    public boolean Permutation(String A, String B) {
        int n = A.length();
        if (n != B.length()) {
            return false;
        }

        int []  count = new int[256];
        for (int i = 0; i < n; i++) {
            ++count[A.charAt(i)];
        }

        for (int i = 0; i < n; i++) {
            if (--count[B.charAt(i)] < 0) {
                return false;
            }
        }

        return true;
    }
};
